pkg/sentry/mm/procfs.go

```diff
@@ -83,6 +85,22 @@ func (mm *MemoryManager) ReadMapsDataInto(ctx context.Context, buf *bytes.Buffer
 	}
 }
 
+// ReadMemFile is called by fs/proc.memDataFile.Read to
+// implement /proc/[pid]/mem.
+func (mm *MemoryManager) ReadMemFile(ctx context.Context, dst usermem.IOSequence, offset int64) (int64, error) {
+	adrr, ok := usermem.Addr(offset).ToRange(uint64(dst.NumBytes()))
+	if !ok {
+		return 0, syserror.EINVAL
+	}
+	ars := usermem.AddrRangeSeqOf(adrr)
+	writer := safemem.FromIOWriter{dst.Writer(ctx)}
```

Comment:
This isn't safe. In the common case, dst.IO is another MemoryManager (specifically, the MM of the calling process). mm.CopyInTo(safemem.FromIOWriter{dst}) will therefore take locks in two MMs, in violation of the lock ordering requirements for usermem.IO. This is extremely deadlock-prone; for example, reading `/proc/self/mem` will deadlock if any other thread attempts to lock MM.activeMu for writing.

Instead, perform an explicitly buffered copy in the implementation of `/proc/[pid]/mem`, e.g.

```
if dst.NumBytes() == 0 {
  return 0, nil
}
buf := make([]byte, dst.NumBytes())
n, rdErr := m.mm().CopyIn(ctx, buf)
if n != 0 {
  _, err := dst.CopyOut(ctx, buf[:n])
  if err != nil {
    // Linux returns EFAULT if fs/proc/base.c:mem_rw() => copy_to_user() fails, forgetting any number of bytes previously copied.
    return 0, syserror.EFAULT
  }
}
if err != nil {
  // Linux converts all errors from fs/proc/base.c:mem_rw() => access_remote_vm() to EIO.
  return n, syserror.EIO
}
return n, nil
```